artemis-core-client/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/ResponseCache.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.activemq.artemis.core.protocol.core.impl;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.apache.activemq.artemis.api.core.ActiveMQInterruptedException;
+import org.apache.activemq.artemis.core.protocol.core.Packet;
+import org.apache.activemq.artemis.core.protocol.core.ResponseHandler;
+
+public class ResponseCache {
+
+   private final AtomicInteger writerPointer = new AtomicInteger(0);
+   private final AtomicInteger sequence = new AtomicInteger(0);
+
+   private final Packet[] store;
+   private ResponseHandler responseHandler;
+
+   public ResponseCache(int size) {
+      this.store = new Packet[size];
+   }
+
+   public long add(Packet packet) {
+      int pointer = writerPointer.getAndUpdate(operand -> {
```

Comment:
The `IntUnaryOperator` must be side effect free (it seems not due to `sequence.incrementAndGet` or am I wrong?) because it may be retried due to contention: [AtomicInteger::getAndUpdate](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicInteger.html#getAndUpdate-java.util.function.IntUnaryOperator-).
If this `ResponseCache` is (or could be turned into) a `single-writer` cache my suggestion is to avoid complex (and expensive) logic here

Reply:
I want to avoid having a lock like bad here, thus this. 
If there is retry done here it would simply set the packet at a new location in the array. And the one that won in the contention would be left set at the other.

Any ideas what to do here instead?

Reply:
It depends on how it is accessed by concurrent writers ie `handleResponse` and `add` could be accessed by different threads concurrently?

Dealing with concurrent accesses here could be dangerous: If multiple threads are accessing concurrently `Packet[]` and the references are not padded by 128 bytes (ie 4 object references between the `Packet` references in the array) the cache won't scale due to false sharing

Reply:
endedup just making a simpler solution using the already existing and tested ConcurrentLongHashMap we have, made all this a lot simpler.